tracing-opentelemetry/src/span_ext.rs

```diff
@@ -42,6 +42,40 @@ pub trait OpenTelemetrySpanExt {
     /// ```
     fn set_parent(&self, cx: Context);
 
+    /// Associates `self` with a given OpenTelemetry trace, using the provided
+    /// followed span [`Context`].
+    ///
+    /// [`Context`]: opentelemetry::Context
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use opentelemetry::{propagation::TextMapPropagator, trace::TraceContextExt};
+    /// use opentelemetry::sdk::propagation::TraceContextPropagator;
+    /// use tracing_opentelemetry::OpenTelemetrySpanExt;
+    /// use std::collections::HashMap;
+    /// use tracing::Span;
+    ///
+    /// // Example carrier, could be a framework header map that impls otel's `Extract`.
+    /// let mut carrier = HashMap::new();
+    ///
+    /// // Propagator can be swapped with b3 propagator, jaeger propagator, etc.
+    /// let propagator = TraceContextPropagator::new();
+    ///
+    /// // Extract otel parent context via the chosen propagator
+    /// let parent_context = propagator.extract(&carrier);
+    ///
+    /// // Generate a tracing span as usual
+    /// let app_root = tracing::span!(tracing::Level::INFO, "app_start");
+    ///
+    /// // Assign parent trace from external context
+    /// app_root.add_link(parent_context.clone());
+    ///
+    /// // Or if the current span has been created elsewhere:
+    /// Span::current().add_link(parent_context);
+    /// ```
+    fn add_link(&self, cx: Context);
```

Comment:
oh, one other thought. I notice that in the examples, the `Context` parameter always has to be cloned...but in the `add_link` method, we borrow the span from the context, and then clone the `SpanContext` from the span...so, afaict, we could just borrow the `Context` parameter rather than forcing the user to clone it when calling this method. that might be more efficient, as well?

but, on the other hand, existing methods on this trait (e.g. `set_parent`) take the `Context` by value, so maybe it's preferable to be consistent? IDK.

@jtescher what do you think?

Reply:
I was trying to be consistent with `set_parent`, as you notice, but that may not be necessary indeed.

Reply:
yeah, i think there's kind of a tradeoff between consistency and ergonomics (and possibly performance?) here...i'd probably go with taking it by reference, but it's @jtescher's call.

alternatively, should this method just take the `SpanContext`? that way, the user can get it from somewhere other than a `Context`...i'm not sure if there's a use-case for that, or not?

Reply:
The method could just take `SpanContext` since links don't care about the whole `Context`.
However propagators return `Context` so maybe we need both `fn add_link(&self, cx: SpanContext)` and `fn add_link(&self, cx: &Context)` .

 `fn add_link(&self, cx: &Context)`  would just get the `SpanContext` of the `Context` and pass it to `fn add_link(&self, cx: SpanContext)`

Reply:
So the [add_link spec](https://github.com/open-telemetry/opentelemetry-specification/blob/2f23c2013ff49cf45de152019367d0ec0f2135d1/specification/trace/api.md#specifying-links) suggests a `SpanContext` directly, while the [span creation spec](https://github.com/open-telemetry/opentelemetry-specification/blob/2f23c2013ff49cf45de152019367d0ec0f2135d1/specification/trace/api.md#span-creation) explicitly requires a full `Context` to set a parent so the parameter discrepancy seems intentional.